docs/internal/dialect.md

```diff
@@ -0,0 +1,116 @@
+
+Verona types have an algebraic semantics and need to be handled carefully by the compiler.
+However, all of the inference and checks are done by the parser, so the MLIR layer can assume the conversions, if any, are correct.
+Therefore, there will be no type checks in MLIR and any conversion between types (ex. a concrete argument to an interface declaration) will be via a `cast` operation that is equivalent to a NO-OP.
+
+Type representation in MLIR is not necessary for correctness.
+With all checks done at parser level and the NO-OP casts at MLIR level, the semantics should not change amidst optimisations.
+However, by keeping the information for longer, we can elide a number of run-time checks at the final lowering, improving performance.
+
+If we decide to move type inference to the MLIR level in the future, this will also help implement the passes, as it is similar to the original representation designed for that.
+
+## Concrete Types
+
+These will be represented by _named_ LLVM `StructureType` pointers.
+Access to fields are done via structure offset, with `embed` fields directly accessed and the rest resolved with a pointer indirection.
+Access to methods will be done by the parser and at this point, dynamic dispatch will already have a `vtable` pointer and an offset.
+Direct (static) calls are done directly as an MLIR call without further complications.
+
+Fields can be reordered for efficiency reasons, for example stronger alignment first to avoid misaligned load/store on later fields.
+This has to be an early static decision, to avoid changing access patterns all over the code later on.
+So it's also likely going to be simple and predictable.
+
+## Interface Types
+
+Abstract interface types that survive to the MLIR layer have their own descriptors, so their own LLVM `StructureType` pointers and types.
+Those are solely used for run-time checks and all access to fields and methods will be done through the run-time concrete object pointer instead.
+
+## Intersection Types
+
+Intersection types will end up as concrete types themselves and treated the same as above.
+For example `(A & B)` will end up as a new `A_&_B` structure with the intersection of both `A` and `B` fields and methods.
+
+Intersection of a concrete or interface type with a capability will be represented specially as `!cap<Type>`.
+
+The four capabilities are:
+* `!iso<Type>` (Isolated)
+* `!mut<Type>` (Mutable)
+* `!imm<Type>` (Immutable)
+* `!ro<Type>` (Read-Only)
+
+This is so that we can carry the capability semantics throughout the MLIR passes and avoid emitting too many run-time checks at final lowering (see below).
+
+## Union Types
+
+Union types will be represented by `!join<A, B>`.
+Verona types are algebraic, so `!join<A, !join<B, C>>` is equivalent to `!join<A, B, C>` and will be represented as the latter in MLIR.
+
+We need to keep union types represented in such a way because this will allow us to simplify calls and checks after optimisations.
```

Comment:
Does Verona normalize intersections of unions to [Disjunctive Normal Form](https://en.wikipedia.org/wiki/Disjunctive_normal_form)?

If so, it might be worth explicitly stating that here.

Reply:
Yes it does! that's a good point, I'll add to the document. Thanks!